content/blog/2022-03-21-conservative-gc/index.md

```diff
@@ -0,0 +1,256 @@
+_ambiguous references_ that could be either pointers or values.
+
+Because exact collectors know references, they can move objects and
+filter out more references. However, exact collectors impose
+non-trivial engineering and performance challenges, such as having to
+maintain a shadow stack. In many situations, it is also impossible to
+implement an exact collector. So, work on implementing effective and
+efficient conservative collectors is necessary.
+
+The key challenges in implementing conservative collectors are
+as follows:
+
+1. **Excess retention**: Because an ambiguous reference _could_ point to
+an object, that "referent" cannot be collected. This is because the
+ambiguous reference _could_ be a valid pointer to that object. So,
+actually dead objects and their transitively reachable descendants
+will be kept alive, causing extra space to be taken.
+
+2. **Pinning**: Because an ambiguous reference _could_ be a value, the
+collector can't modify it. So, in the case that the ambiguous
+reference is actually a pointer, the referents cannot be moved,
+causing fragmentation.
+
+# Contributions
+
+The main contributions of this paper include:
+
+- A detailed examination on conservative garbage collection, including
+  the first detailed study of impacts of exact and conservative
+  collection in practice.
+
+- Introducing an object map that precisely tracks live objects to filter
+  ambiguous roots efficiently.
+
+- The design, implementation, and evaluation of new conservative
+  collectors that use the object map, against prior conservative
+  collectors and their exact counterparts.
+
+  - **RC Immix<sub>Cons</sub>**, introduced in this paper, is 1%
+  _faster_ than **Gen Immix**, the best-performing exact collector.
+
+<center>
+<img src="fig1.png" alt="Exact versus Conservative Collectors" width="50%" />
+</center>
+
+# Techniques
+
+There are two broad classes of techniques used by conservative collectors to
+make them match exact collectors:
+
+1. **Filtering**: maintaining a data structure to track references to live objects
+   to filter ambiguous references that could not possibly be valid pointers to reduce
+   excess retention.
+
+2. **Ambiguous Roots**: many collectors strike the balance between exact and conservative
+   collection by using available run-time information such that the only conservative references
+   are the roots. Other references are exactly known, and therefore can be moved freely.
+
+## Existing Conservative Collectors
+
+**Boehm-Demers-Weiser Collector (BDW)**[^2] is a conservative collector designed for C and C++. It is a
+non-moving tracing collector which uses a free-list to reclaim and reuse memory which can be used
+as a drop-in replacement for `malloc`. It filters ambiguous references by comparing them against
+live free-list cells. While it is very convenient to use, it suffers from bad mutator performance
+due to sparse allocations and the inability to defragment.
```

Comment:
Maybe worth pointing out here that, in the real world, this collector is far and away the most popular one (because it is so easy to bolt onto your code without changing a thing). Like, just the other day, a [post was on Hacker News](https://acha.ninja/blog/memory-safeish-hare/) describing yet another use of the Boehm collector.

Reply:
Linked the article as an example.